File: golang-http-server/internal/service/project_service.go
package service

import (
	"context"

	"golang-http-server/internal/core"
	"golang-http-server/internal/core/errors"
	"golang-http-server/internal/db/models"
	"golang-http-server/internal/dto"
	"golang-http-server/internal/repository"
)

type ProjectService struct {
	repo *repository.ProjectRepository
}

func NewProjectService(repo *repository.ProjectRepository) *ProjectService {
	return &ProjectService{repo: repo}
}

func (s *ProjectService) List(ctx context.Context, search, status *string, limit, offset int) ([]dto.ProjectResponse, core.PaginationMeta, error) {
	logger := core.GetLogger()
	logger.Debug("listing projects", "search", search, "status", status, "limit", limit, "offset", offset)

	projects, total, err := s.repo.List(ctx, search, status, limit, offset)
	if err != nil {
		return nil, core.PaginationMeta{}, err
	}

	meta := core.BuildPaginationMeta(total, limit, offset)
	responses := make([]dto.ProjectResponse, len(projects))
	for i, p := range projects {
		responses[i] = projectToResponse(p)
	}

	return responses, meta, nil
}

func (s *ProjectService) FindByID(ctx context.Context, id string) (*dto.ProjectResponse, error) {
	logger := core.GetLogger()
	logger.Debug("finding project by ID", "id", id)

	project, err := s.repo.GetByID(ctx, id)
	if err != nil {
		return nil, err
	}
	if project == nil {
		return nil, errors.NewNotFoundError("Project")
	}

	response := projectToResponse(*project)
	return &response, nil
}

func (s *ProjectService) Create(ctx context.Context, params dto.ProjectCreate, ownerID *string) (*dto.ProjectResponse, error) {
	logger := core.GetLogger()
	logger.Info("creating project", "name", params.Name)

	if err := params.Validate(); err != nil {
		return nil, err
	}

	project, err := s.repo.Create(ctx, params.Name, params.Description, params.Status, ownerID)
	if err != nil {
		return nil, err
	}

	response := projectToResponse(*project)
	return &response, nil
}

func (s *ProjectService) Update(ctx context.Context, id string, params dto.ProjectUpdate) (*dto.ProjectResponse, error) {
	logger := core.GetLogger()
	logger.Info("updating project", "id", id)

	if err := params.Validate(); err != nil {
		return nil, err
	}

	project, err := s.repo.Update(ctx, id, params.Name, params.Description, params.Status)
	if err != nil {
		return nil, err
	}
	if project == nil {
		return nil, errors.NewNotFoundError("Project")
	}

	response := projectToResponse(*project)
	return &response, nil
}

func (s *ProjectService) Delete(ctx context.Context, id string) error {
	logger := core.GetLogger()
	logger.Info("deleting project", "id", id)

	exists, err := s.repo.Exists(ctx, id)
	if err != nil {
		return err
	}
	if !exists {
		return errors.NewNotFoundError("Project")
	}

	return s.repo.Delete(ctx, id)
}

func (s *ProjectService) HealthCheck(ctx context.Context) bool {
	count, err := s.repo.Count(ctx, "")
	return err == nil && count >= 0
}

func projectToResponse(p models.Project) dto.ProjectResponse {
	resp := dto.ProjectResponse{
		ID:        p.ID,
		Name:      p.Name,
		Status:    p.Status,
		CreatedAt: p.CreatedAt,
		UpdatedAt: p.UpdatedAt,
	}
	if p.Description.Valid {
		resp.Description = &p.Description.String
	}
	if p.OwnerID.Valid {
		resp.OwnerID = &p.OwnerID.String
	}
	return resp
}

File: fastapi/src/services/project_service.py
from uuid import UUID

from fastapi import Depends
from sqlalchemy.ext.asyncio import AsyncSession

from src.core.errors.not_found_error import NotFoundError
from src.core.logger import get_logger
from src.core.pagination import PaginationMeta, build_pagination_meta
from src.db import get_db
from src.db.models.project import Project
from src.repositories.project_repository import ProjectRepository, get_project_repository
from src.schemas.project_schemas import ProjectCreate, ProjectUpdate


class ProjectService:
    def __init__(self, repository: ProjectRepository) -> None:
        self._repository = repository

    async def list(
        self,
        *,
        search: str | None = None,
        status: str | None = None,
        limit: int = 20,
        offset: int = 0,
    ) -> tuple[list[Project], PaginationMeta]:
        logger = get_logger()
        logger.debug("Listing projects", search=search, status=status, limit=limit, offset=offset)

        items, total = await self._repository.list_with_filters(
            search=search, status=status, limit=limit, offset=offset
        )

        meta = build_pagination_meta(total, limit, offset)
        return items, meta

    async def find_by_id(self, id: UUID) -> Project:
        logger = get_logger()
        logger.debug("Finding project by ID", id=str(id))

        project = await self._repository.get_by_id(id)
        if project is None:
            raise NotFoundError("Project")
        return project

    async def create(self, params: ProjectCreate, owner_id: UUID | None = None) -> Project:
        logger = get_logger()
        logger.info("Creating project", name=params.name)

        return await self._repository.create(
            name=params.name,
            description=params.description,
            status=params.status,
            owner_id=owner_id,
        )

    async def update(self, id: UUID, params: ProjectUpdate) -> Project:
        logger = get_logger()
        logger.info("Updating project", id=str(id))

        project = await self.find_by_id(id)

        updates: dict[str, str | None] = {}
        if params.name is not None:
            updates["name"] = params.name
        if params.description is not None:
            updates["description"] = params.description
        if params.status is not None:
            updates["status"] = params.status

        return await self._repository.update(project, **updates)

    async def delete(self, id: UUID) -> bool:
        logger = get_logger()
        logger.info("Deleting project", id=str(id))

        project = await self.find_by_id(id)
        await self._repository.delete(project)
        return True

    async def health_check(self) -> bool:
        try:
            await self._repository.count()
            return True
        except Exception:
            return False


def get_project_service(session: AsyncSession = Depends(get_db)) -> ProjectService:
    repository = get_project_repository(session)
    return ProjectService(repository)
